### ISSUE TITLE
brew install lastpass-cli --with-pinentry fails when XCode CLI installed but not XCode itself

### ISSUE THREAD

**Issue by cjcjameson** [lastpass/lastpass-cli#133]
Hi there, just letting you know of the following:

If you have the XCode CLI tools installed, but not the XCode program itself, `brew install lastpass-cli --with-pinentry` fails:

```
brew install lastpass-cli --with-pinentry
==> Downloading https://github.com/lastpass/lastpass-cli/archive/v0.7.0.tar.gz
Already downloaded: /Library/Caches/Homebrew/lastpass-cli-0.7.0.tar.gz
==> make PREFIX=/usr/local/Cellar/lastpass-cli/0.7.0 install
Last 15 lines from /Users/swag/Library/Logs/Homebrew/lastpass-cli/01.make:
In file included from cipher.c:36:
In file included from ./cipher.h:4:
./kdf.h:4:10: fatal error: 'sys/types.h' file not found
#include <sys/types.h>
         ^
1 error generated.
make: *** [blob.o] Error 1
In file included from clipboard.c:38:
./util.h:4:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^
1 error generated.
make: *** [clipboard.o] Error 1
1 error generated.
make: *** [cipher.o] Error 1

READ THIS: https://git.io/brew-troubleshooting
```

Installing XCode itself resolved the issue and unblocked brew. I'm on a pretty fresh install of El Capitan.

Brew's unblocked, although `make` fails; I'm moving along but looks like there's an openssl dependency I'm missing that wasn't automatically installed or checked for:

```
cc -O3 -march=native -fomit-frame-pointer -pipe -std=gnu99 -D_GNU_SOURCE -pedantic -Wall -Wextra -Wno-language-extension-token -MMD -Wno-deprecated-declarations -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/usr/include/libxml2   -c -o lpass.o lpass.c
In file included from lpass.c:37:
In file included from ./cmd.h:4:
In file included from ./blob.h:4:
./kdf.h:5:10: fatal error: 'openssl/sha.h' file not found
#include <openssl/sha.h>
         ^
1 error generated.
make: *** [lpass.o] Error 1
```

**Comment by bcopeland** [comment]
I'll  admit to not knowing much about the brew recipes and the vagaries of XCode -- so if someone on a mac can chime in or send a PR that would be great.

**Comment by gnawhleinad** [comment]
> although make fails

It fails because it can't find the `openssl` library (and also `libxml2`). You can use:

```
$ make CFLAGS="-I$(brew --prefix)/opt/openssl/include -I$(brew --prefix)/opt/libxml2/include/libxml2"
```

to build `lpass` on OS X. Alternatively, you can use brew to force `link` the libraries.

> Installing XCode itself resolved the issue and unblocked brew.

I'm not certain, but this isn't a problem of XCode vs. XCode CLI but a matter of having all the [dependencies](https://github.com/lastpass/lastpass-cli#dependencies) installed.

**Comment by bcopeland** [comment]
@cjcjameson can you confirm that just having openssl etc installed fixes it for you?

**Comment by cjcjameson** [comment]
Thanks y'all. Alas, I can't reproduce it exactly, but I do have openssl
installed and it is working now.

On Wed, Dec 2, 2015 at 1:10 PM, Bob Copeland <notifications@github.com>
wrote:

> @cjcjameson <https://github.com/cjcjameson> can you confirm that just
> having openssl etc installed fixes it for you?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lastpass/lastpass-cli/issues/133#issuecomment-161434932>
> .
>



-- 
C.J. Jameson
Pivotal Labs

**Comment by travi** [comment]
I'm having the same issue as the first problem mentioned above. I only have `xcode-select` installed and would prefer not to upgrade to the full blown Xcode since the lighter weight command-line tools cover all of my other needs. 

Any chance this could be updated to not have a dependency on the full Xcode?

**Comment by bcopeland** [comment]
The listed dependency is stdio.h which is in libc; I don't see how that particular dependency could go away.

But I know nothing about the various Xcode flavors and which gets you libc or not.  Perhaps asking the brew developers would get you a better answer?

**Comment by travi** [comment]
Fair enough. Thanks for the response.

**Comment by kalbasit** [comment]
@travi if you do open an issue with Homebrew, please CC me as well. Having the same issue.

**Comment by bcopeland** [comment]
BTW the brew recipe for lastpass is over here:
https://github.com/Homebrew/homebrew/blob/master/Library/Formula/lastpass-cli.rb

If you do find something we should add there, please do send a PR to brew with a fix.

**Comment by travi** [comment]
Admittedly, I'm not all that familiar with how brew formulas work either. Plus, I primarily use lastpass from the browser, so having a broken cli isn't a top priority to contribute to getting it fixed. I'm happy to raise visibility, but since I wasn't the one that determined that the issue could be related to Xcode flavors, etc, I probably won't be much help when they ask for deeper details.

If one of you guys get to logging the issue before I get around to it, it will probably be better for getting it actually resolved. Please do mention it in this thread if you open something. Otherwise, I'll keep it on my list to follow up on at some point.

**Comment by kalbasit** [comment]
It's worth mentioning that this has started failing for me recently. I did install lastpass-cli version 0.6.0 but it's failing to update to v0.7.0

```
kalbasit@cratos ~/.dotfiles [master *] ± % brew info lastpass-cli
lastpass-cli: stable 0.7.0 (bottled), HEAD
LastPass command-line interface tool
https://github.com/lastpass/lastpass-cli
/usr/local/Cellar/lastpass-cli/0.6.0 (5 files, 144K) *
  Built from source with: --with-doc
From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/lastpass-cli.rb
==> Dependencies
Required: openssl ✔
Optional: pinentry ✘
==> Options
--with-doc
        Install man pages
--with-pinentry
        Build with pinentry support
--HEAD
        Install HEAD version
kalbasit@cratos ~/.dotfiles [master *] ± % brew upgrade lastpass-cli
==> Upgrading 1 outdated package, with result:
lastpass-cli 0.7.0
==> Upgrading lastpass-cli
==> Downloading https://github.com/lastpass/lastpass-cli/archive/v0.7.0.tar.gz
Already downloaded: /Library/Caches/Homebrew/lastpass-cli-0.7.0.tar.gz
==> make PREFIX=/usr/local/Cellar/lastpass-cli/0.7.0 install
Last 15 lines from /Users/kalbasit/Library/Logs/Homebrew/lastpass-cli/01.make:
./util.h:4:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^
1 error generated.
make: *** [clipboard.o] Error 1
In file included from cmd-duplicate.c:36:
In file included from ./cmd.h:4:
In file included from ./blob.h:4:
./kdf.h:4:10: fatal error: 'sys/types.h' file not found
#include <sys/types.h>
         ^
1 error generated.
make: *** [cmd-duplicate.o] Error 1
1 error generated.
make: *** [cipher.o] Error 1

READ THIS: https://git.io/brew-troubleshooting
```

**Comment by cjcjameson** [comment]
from @d, he says:
```
After closer inspection it seems that the --with-pinentry is causing the local compilation, can we get rid of that flag? I'm personally a fan of pinentry when on Linux, but the Mac OS port is curses based, which means that it brings zero value to the security...
```

**Comment by kalbasit** [comment]
Mine is failing without `--with-pinentry`. Look at the logs above.

**Comment by kalbasit** [comment]
@bcopeland this might be due to some Makefile magic. See https://github.com/Homebrew/homebrew/issues/46763#issuecomment-162708655

**Comment by bcopeland** [comment]
Nice.  So, is there a way to tell SDKROOT without Xcode installed?

**Comment by kalbasit** [comment]
@bcopeland I'm not all familiar with Mac compiler either, mind asking @apjanke on https://github.com/Homebrew/homebrew/issues/46763#issuecomment-162708655 ?

**Comment by apjanke** [comment]
I'm still learning this stuff, but here's my current understanding: I don't think there's a way of setting SDKROOT without Xcode installed per se, but if you have the Command Line Tools installed, then you probably don't need to set an SDKROOT, because the headers and libs are dropped in the traditional Unix locations under `/usr/include`, `/usr/lib`, and so on. Using the non-Darwin code with `pkg-config` [in the `else` block](https://github.com/lastpass/lastpass-cli/blob/39b6c05500dbe26c439ca44dd989bd20df7208c2/Makefile#L17-L22) should work in that scenario.

I'm afraid I don't know the graceful way of detecting whether CLT are installed. But if they aren't installed, you'll just get a failed build and a dialog asking the user if they want to install the CLT. So you could probably just do that `xcodebuild -sdk` test, and if it tells you it's not pointing to Xcode, just fall back to doing the stuff in the `else` block. Generally, once the CLT are installed, you can treat OS X as a generic Unix/BSD system, and not have to do special Xcode stuff.

Happy to fiddle with the Homebrew formula to get it playing nicely with your build system. I think I'm going to try hacking it to just run that `else` block in a CLT-only environment and see how it goes.

**Comment by kalbasit** [comment]
@apjanke @bcopeland any news on this?

**Comment by apjanke** [comment]
Sorry, haven't had time to try it yet.

**Comment by bcopeland** [comment]
Ditto, I'm actually waiting on some hardware to get an osx env setup.

**Comment by kalbasit** [comment]
@bcopeland @apjanke OK so I took some time to try your suggestion and you are right. If we default to the `else` block it should work.

WARNING: I suck at Makefile so please re-do it as you see fit and I can test it for you.

Homebrew patch (which in turns patches the Makefile)
```
diff --git a/Library/Formula/lastpass-cli.rb b/Library/Formula/lastpass-cli.rb
index 6f808c6..8d558a9 100644
--- a/Library/Formula/lastpass-cli.rb
+++ b/Library/Formula/lastpass-cli.rb
@@ -18,6 +18,9 @@ class LastpassCli < Formula
   depends_on "openssl"
   depends_on "pinentry" => :optional

+  # Fix Makefile
+  patch :DATA
+
   def install
     system "make", "PREFIX=#{prefix}", "install"
     system "make", "MANDIR=#{man}", "install-doc" if build.with? "doc"
@@ -27,3 +30,22 @@ class LastpassCli < Formula
     system "#{bin}/lpass", "--version"
   end
 end
+
+__END__
+diff -Naur a/Makefile b/Makefile
+--- a/Makefile  2015-12-14 20:44:21.000000000 -0800
++++ b/Makefile  2015-12-14 20:46:58.000000000 -0800
+@@ -12,8 +12,13 @@
+ UNAME_S := $(shell sh -c 'uname -s 2>/dev/null || echo not')
+ ifeq ($(UNAME_S),Darwin)
+ SDKROOT ?= $(shell xcodebuild -version -sdk macosx | sed -n 's/^Path: \(.*\)/\1/p')
++ifeq ($(SDKROOT),)
++CFLAGS += $(shell pkg-config --cflags libxml-2.0 2>/dev/null || echo -I/usr/include/libxml2) -I/usr/local/include
++LDLIBS = -lcurl $(shell pkg-config --libs libxml-2.0 2>/dev/null || echo -lxml2) -lssl -lcrypto
++else
+ CFLAGS += -Wno-deprecated-declarations -isysroot $(SDKROOT) -I$(SDKROOT)/usr/include/libxml2
+ LDLIBS = -lcurl -lxml2 -lssl -lcrypto
++endif
+ else
+ CFLAGS += $(shell pkg-config --cflags libxml-2.0 2>/dev/null || echo -I/usr/include/libxml2) -I/usr/local/include
+ LDLIBS = -lcurl $(shell pkg-config --libs libxml-2.0 2>/dev/null || echo -lxml2) -lssl -lcrypto
```

**Comment by bcopeland** [comment]
Please test with latest master, I think it works now (at least I tried the CLT version and it worked OK for me).

**Comment by kalbasit** [comment]
nope :(

```
$ brew update
Updated Homebrew from 1df4379b to b016eaa6.
Updated 5 taps (caskroom/cask, caskroom/versions, homebrew/bundle, homebrew/dupes, homebrew/games).
==> New Formulae
cern-ndiff                    codec2                        gjs                           gnome-builder                 homebrew/games/fifechan       mydumper                      qriollo                       textql                        vamp-plugin-sdk
==> Updated Formulae
afl-fuzz                aws-elasticbeanstalk ✔  carina                  flow                    ircd-hybrid             libcroco                lorem                   node ✔                  rancher-compose         sip                     wget ✔
ammonite-repl           awscli ✔                cgal                    ghq                     jruby                   libstrophe              macvim                  open-mpi                ruby                    solr                    yank
apt-cacher-ng           bind                    charm-tools             giter8                  juju-quickstart         libtorrent-rasterbar    mpd                     openvpn                 ruby-build              texi2html               youtube-dl ✔
arping                  burp                    elasticsearch           glide                   kawa                    libvisio                mpich                   osquery                 scamper                 tomcat-native           zeromq
avrdude                 caddy                   erlang ✔                global                  kotlin-compiler         lighttpd                nanomsg                 qbs                     sec                     vert.x
$ brew install --with-doc lastpass-cli
==> Downloading https://github.com/lastpass/lastpass-cli/archive/v0.7.0.tar.gz
Already downloaded: /Library/Caches/Homebrew/lastpass-cli-0.7.0.tar.gz
==> make PREFIX=/usr/local/Cellar/lastpass-cli/0.7.0 install
Last 15 lines from /Users/kalbasit/Library/Logs/Homebrew/lastpass-cli/01.make:
./kdf.h:4:10: fatal error: 'sys/types.h' file not found
#include <sys/types.h>
         ^
1 error generated.
make: *** [cmd-edit.o] Error 1
In file included from cmd-export.c:36:
In file included from ./cmd.h:4:
In file included from ./blob.h:4:
./kdf.h:4:10: fatal error: 'sys/types.h' file not found
#include <sys/types.h>
         ^
1 error generated.
make: *** [cmd-export.o] Error 1
1 error generated.
make: *** [cipher.o] Error 1

READ THIS: https://git.io/brew-troubleshooting

These open issues may also help:
lastpass-cli v0.7.0 fails to build https://github.com/Homebrew/homebrew/issues/46763
```

**Comment by bcopeland** [comment]
On Fri, Dec 18, 2015 at 09:33:49AM -0800, Wael M. Nasreddine wrote:
> nope :(
> ==> Downloading https://github.com/lastpass/lastpass-cli/archive/v0.7.0.tar.gz
> Already downloaded: /Library/Caches/Homebrew/lastpass-cli-0.7.0.tar.gz

This is pulling the last release though, not the HEAD of the master
branch.

-- 
Bob Copeland %% http://bobcopeland.com/

**Comment by apjanke** [comment]
Yep!

You can use `brew install --HEAD <formula>` to install it from the head of master.

This is on my 10.9.5 box which has only CLT installed, no Xcode.

Version 0.7.0:
```
$ brew install lastpass-cli --with-pinentry
==> Using the sandbox
==> Downloading https://github.com/lastpass/lastpass-cli/archive/v0.7.0.tar.gz
==> Downloading from https://codeload.github.com/lastpass/lastpass-cli/tar.gz/v0
######################################################################## 100.0%
==> make PREFIX=/usr/local/Cellar/lastpass-cli/0.7.0 install
Last 15 lines from /Users/janke/Library/Logs/Homebrew/lastpass-cli/01.make:
make: *** [lpass.o] Error 1
In file included from blob.c:36:
In file included from ./blob.h:4:
./kdf.h:4:10: fatal error: 'sys/types.h' file not found
#include <sys/types.h>
         ^
1 error generated.
make: *** [blob.o] Error 1
In file included from cipher.c:36:
In file included from ./cipher.h:4:
./kdf.h:4:10: fatal error: 'sys/types.h' file not found
#include <sys/types.h>
         ^
1 error generated.
make: *** [cipher.o] Error 1
```


HEAD:

```
[~]
$ brew install lastpass-cli --HEAD --with-pinentry
==> Using the sandbox
==> Cloning https://github.com/lastpass/lastpass-cli.git
Updating /Library/Caches/Homebrew/lastpass-cli--git
==> Checking out branch master
==> make PREFIX=/usr/local/Cellar/lastpass-cli/HEAD install
🍺  /usr/local/Cellar/lastpass-cli/HEAD: 4 files, 164K, built in 24 seconds
[~]
$
```

The Makefile change looks good to me.

What's your release cycle like? If you cut a 0.7.1 release some time soon, we can just bump Homebrew to grab that, and not worry about patches.

**Comment by bcopeland** [comment]
On Fri, Dec 18, 2015 at 01:18:33PM -0800, Andrew Janke wrote:
> What's your release cycle like? If you cut a 0.7.1 release some time soon, we can just bump Homebrew to grab that, and not worry about patches.

I'll try to cut a new release early next week.

-- 
Bob Copeland %% http://bobcopeland.com/

**Comment by bcopeland** [comment]
Okay, I released a 0.7.1 with just that patch on top of 0.7.0.  The HEAD needs just a little bit more baking before 0.8.0 is ready.
